### PR TITLE
Telemetry Ongoing Implementation

### PR DESCRIPTION
**Pull request by rayluo**
This is NOT YET a PR ready for merge. We use it to share the progress and facilitate the discussion.

The first commit contains only the public API. There are 2 topics open for discussion, annotated as `TBD` inside the source code already.

### CODE REVIEW

File: src/Microsoft.Identity.Client/Telemetry.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+//
+//------------------------------------------------------------------------------
+
+namespace Microsoft.Identity.Client
+{
+    public class Telemetry
+    {
+        private static readonly Telemetry Singleton = new Telemetry();
+
+        private IReceiver Receiver = null;  // TBD: Actually we could just use a delegate EventAction(List<Dictionary<string, string>> events) here, not a full class.
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Telemetry.cs]: @addev-ashish - can this be a delegate instead for .NET? seems more natural.

File: src/Microsoft.Identity.Client/IReceiver.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+//
+//------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+
+namespace Microsoft.Identity.Client
+{
+    public interface IReceiver
+    {
+        void onEvents(List<Dictionary<string, string>> events); // Renamed from Dispatcher.dispatchEvent(...)
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/IReceiver.cs]: method name starts with upper case

File: tests/Test.MSAL.NET.Unit/TelemetryTests.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    [TestClass]
+    public class TelemetryTests
+    {
+        [TestInitialize]
+        public void Initialize()
+        {
+        }
+
+        [TestMethod]
```
Comment (kpanwar) [inline tests/Test.MSAL.NET.Unit/TelemetryTests.cs]: add category for tests like other classes

File: tests/Test.MSAL.NET.Unit/TelemetryTests.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        public void TelemetryPublicApiSample()
+        {
+            Telemetry telemetry = Telemetry.GetInstance();
```
Comment (kpanwar) [inline tests/Test.MSAL.NET.Unit/TelemetryTests.cs]: assert that telemetry is not null.

File: tests/Test.MSAL.NET.Unit/TelemetryTests.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            MyReceiver receiver = new MyReceiver();
+            telemetry.RegisterReceiver(receiver);
+
+            // Or you can use a one-liner:
+            Telemetry.GetInstance().RegisterReceiver(new MyReceiver());
```
Comment (kpanwar) [inline tests/Test.MSAL.NET.Unit/TelemetryTests.cs]: add another assert to check that getInstance returns same object for each call. you can use object.Equals(obj);

File: src/Microsoft.Identity.Client/IReceiver.cs (hunk 2 above)
Comment (addev-ashish) [inline src/Microsoft.Identity.Client/IReceiver.cs]: remove the comment about the rename.

---
In reply to: [108025281](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/264#discussion_r108025281) [](ancestors = 108025281)

File: src/Microsoft.Identity.Client/IReceiver.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        /*
+         * TBD: Open Discussion:
+         * Will the values of collected events always be string?
```
Comment (addev-ashish) [inline src/Microsoft.Identity.Client/IReceiver.cs]: >Will the values of collected events always be string [](start = 11, length = 52)

yes

File: src/Microsoft.Identity.Client/IReceiver.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+         * How do we plan to handle numeric metrics, such as start_time and end_time and api_id?
```
Comment (addev-ashish) [inline src/Microsoft.Identity.Client/IReceiver.cs]: >How do we plan to handle numeric metrics, such as start_time and end_time and api_i [](start = 11, length = 83)

convert to string

File: src/Microsoft.Identity.Client/IReceiver.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+         * And do we need to handle booleans?
```
Comment (addev-ashish) [inline src/Microsoft.Identity.Client/IReceiver.cs]: >And do we need to handle booleans [](start = 11, length = 33)

convert to string true false

File: src/Microsoft.Identity.Client/Telemetry.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+namespace Microsoft.Identity.Client
+{
+    public class Telemetry
```
Comment (addev-ashish) [inline src/Microsoft.Identity.Client/Telemetry.cs]: do you need a private constructor?

File: src/Microsoft.Identity.Client/Telemetry.cs (hunk 1 above)
Comment (addev-ashish) [inline src/Microsoft.Identity.Client/Telemetry.cs]: i think it can be a delegate, it just looks like a fancy setter to me.

---
In reply to: [108025263](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/264#discussion_r108025263) [](ancestors = 108025263)

Comment (rayluo) [inline src/Microsoft.Identity.Client/Telemetry.cs]: Not necessarily a setter. We could still use the RegisterReceiver method if we prefer. It is just that the parameter of that method will become a delegate, rather than an interface.

File: tests/Test.MSAL.NET.Unit/TelemetryTests.cs (hunk 4 above)
Comment (rayluo) [inline tests/Test.MSAL.NET.Unit/TelemetryTests.cs]: I would suggest to keep this `TelemetryPublicApiSample` as-is, in order to demonstrate how to use the Telemetry, without any noise. I would add a separated test case to assert telemetry is not null and assert it is a singleton.

File: tests/Test.MSAL.NET.Unit/TelemetryTests.cs (hunk 5 above)
Comment (rayluo) [inline tests/Test.MSAL.NET.Unit/TelemetryTests.cs]: I would suggest to keep this `TelemetryPublicApiSample` as-is, in order to demonstrate how to use the Telemetry, without any noise. I would add a separated test case to assert telemetry is not null and assert it is a singleton.

File: src/Microsoft.Identity.Client/Telemetry.cs (hunk 1 above)
Comment (addev-ashish) [inline src/Microsoft.Identity.Client/Telemetry.cs]: @rayluo I am ok with it, if it makes sense for .net.

File: tests/Test.MSAL.NET.Unit/TelemetryTests.cs (hunk 3 above)
Comment (rayluo) [inline tests/Test.MSAL.NET.Unit/TelemetryTests.cs]: Done. See the next line. vvv
